### PR TITLE
Network status tracking

### PR DESCRIPTION
**Pull request by makdharma**
Recreated the old PR (https://github.com/grpc/grpc/pull/6283) in a new
repo since the old repo was destroyed. Removed changes to tcp_posix.c
and tpc_windows.c, instead depending on the idempotent endpoint shutdown code
from https://github.com/grpc/grpc/pull/6911.

### CODE REVIEW

**Comment by makdharma** [comment]
@nicolasnoble any idea why the nework_status_change test comes up as failing?

File: src/core/lib/iomgr/network_status_tracker.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// other threads might be in the process of shutdown as well, but that has
+// no side effect.
+void grpc_network_status_shutdown_all_endpoints() {
+  if (head == NULL) {
+    return;
+  }
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+
+  for (endpoint_ll_node *curr = head; curr != NULL; curr = curr->next) {
+    curr->ep->vtable->shutdown(&exec_ctx, curr->ep);
```
Comment (makdharma) [inline src/core/lib/iomgr/network_status_tracker.c]: @ctiller - In failing tests the vtable value is getting corrupt(?) to 0xffffffffffffffff, and the shutdown call leads to a segfault. What could be causing the vtable field to change? A casual inspection of code didn't turn up any place where vtable is getting invalidated.

Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: ASAN output tells a story: https://grpc-testing.appspot.com/job/gRPC_pull_requests/10071/config=asan,language=c,platform=linux/consoleFull

```
=================================================================
==21487==ERROR: AddressSanitizer: heap-use-after-free on address 0x61400000fe40 at pc 0x0000005b810f bp 0x7ffe4ee84730 sp 0x7ffe4ee84728
READ of size 8 at 0x61400000fe40 thread T0
    #0 0x5b810e in grpc_network_status_shutdown_all_endpoints /var/local/git/grpc/src/core/lib/iomgr/network_status_tracker.c:79:15
    #1 0x54cf18 in test_invoke_network_status_change /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:190:3
    #2 0x54b9b2 in network_status_change /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:240:3
    #3 0x4f56ee in grpc_end2end_tests /var/local/git/grpc/test/core/end2end/end2end_nosec_tests.c:308:7
    #4 0x4f38d3 in main /var/local/git/grpc/test/core/end2end/fixtures/h2_census.c:127:5
    #5 0x7f94c507ab44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)
    #6 0x41c005 in _start (/var/local/git/grpc/bins/asan/h2_census_nosec_test+0x41c005)

0x61400000fe40 is located 0 bytes inside of 408-byte region [0x61400000fe40,0x61400000ffd8)
freed by thread T0 here:
    #0 0x4c1a60 in __interceptor_free /llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:38
    #1 0x6a7368 in gpr_free /var/local/git/grpc/src/core/lib/support/alloc.c:68:3
    #2 0x5c6743 in tcp_free /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.c:120:3
    #3 0x5c1b3d in tcp_unref /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.c:148:5
    #4 0x5c28f7 in tcp_destroy /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.c:157:3
    #5 0x5ab6e2 in grpc_endpoint_destroy /var/local/git/grpc/src/core/lib/iomgr/endpoint.c:62:3
    #6 0x5a2f39 in finish /var/local/git/grpc/src/core/lib/http/httpcli.c:112:5
    #7 0x5a298f in on_read /var/local/git/grpc/src/core/lib/http/httpcli.c:162:5
    #8 0x5b4886 in grpc_exec_ctx_flush /var/local/git/grpc/src/core/lib/iomgr/exec_ctx.c:70:7
    #9 0x73e021 in pollset_work /var/local/git/grpc/src/core/lib/iomgr/ev_poll_posix.c:989:22
    #10 0x5b36f3 in grpc_pollset_work /var/local/git/grpc/src/core/lib/iomgr/ev_posix.c:197:10
    #11 0x585b12 in grpc_pick_port_using_server /var/local/git/grpc/test/core/util/port_server_client.c:221:47
    #12 0x581672 in grpc_pick_unused_port /var/local/git/grpc/test/core/util/port_posix.c:188:16
    #13 0x58307f in grpc_pick_unused_port_or_die /var/local/git/grpc/test/core/util/port_posix.c:233:14
    #14 0x4f3a53 in chttp2_create_fixture_fullstack /var/local/git/grpc/test/core/end2end/fixtures/h2_census.c:61:14
    #15 0x54dea9 in begin_test /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:59:7
    #16 0x54bd7a in test_invoke_network_status_change /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:112:7
    #17 0x54b9b2 in network_status_change /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:240:3
    #18 0x4f56ee in grpc_end2end_tests /var/local/git/grpc/test/core/end2end/end2end_nosec_tests.c:308:7
    #19 0x4f38d3 in main /var/local/git/grpc/test/core/end2end/fixtures/h2_census.c:127:5
    #20 0x7f94c507ab44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)

previously allocated by thread T0 here:
    #0 0x4c1d68 in malloc /llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:52
    #1 0x6a7217 in gpr_malloc /var/local/git/grpc/src/core/lib/support/alloc.c:58:7
    #2 0x5c0a2b in grpc_tcp_create /var/local/git/grpc/src/core/lib/iomgr/tcp_posix.c:458:31
    #3 0x5c012b in on_writable /var/local/git/grpc/src/core/lib/iomgr/tcp_client_posix.c:193:13
    #4 0x5b4886 in grpc_exec_ctx_flush /var/local/git/grpc/src/core/lib/iomgr/exec_ctx.c:70:7
    #5 0x73e021 in pollset_work /var/local/git/grpc/src/core/lib/iomgr/ev_poll_posix.c:989:22
    #6 0x5b36f3 in grpc_pollset_work /var/local/git/grpc/src/core/lib/iomgr/ev_posix.c:197:10
    #7 0x585b12 in grpc_pick_port_using_server /var/local/git/grpc/test/core/util/port_server_client.c:221:47
    #8 0x581672 in grpc_pick_unused_port /var/local/git/grpc/test/core/util/port_posix.c:188:16
    #9 0x58307f in grpc_pick_unused_port_or_die /var/local/git/grpc/test/core/util/port_posix.c:233:14
    #10 0x4f3a53 in chttp2_create_fixture_fullstack /var/local/git/grpc/test/core/end2end/fixtures/h2_census.c:61:14
    #11 0x54dea9 in begin_test /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:59:7
    #12 0x54bd7a in test_invoke_network_status_change /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:112:7
    #13 0x54b9b2 in network_status_change /var/local/git/grpc/test/core/end2end/tests/network_status_change.c:240:3
    #14 0x4f56ee in grpc_end2end_tests /var/local/git/grpc/test/core/end2end/end2end_nosec_tests.c:308:7
    #15 0x4f38d3 in main /var/local/git/grpc/test/core/end2end/fixtures/h2_census.c:127:5
    #16 0x7f94c507ab44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)

SUMMARY: AddressSanitizer: heap-use-after-free /var/local/git/grpc/src/core/lib/iomgr/network_status_tracker.c:79:15 in grpc_network_status_shutdown_all_endpoints
Shadow bytes around the buggy address:
  0x0c287fff9f70: 00 00 00 00 00 00 00 00 00 00 00 fa fa fa fa fa
  0x0c287fff9f80: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c287fff9f90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c287fff9fa0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c287fff9fb0: 00 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa
=>0x0c287fff9fc0: fa fa fa fa fa fa fa fa[fd]fd fd fd fd fd fd fd
  0x0c287fff9fd0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c287fff9fe0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c287fff9ff0: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c287fffa000: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c287fffa010: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==21487==ABORTING
```

Comment (murgatroid99) [inline src/core/lib/iomgr/network_status_tracker.c]: That sounds similar to an issue I had once with Ruby: call pointers were becoming invalid while corresponding batches were still pending, which led to what looked like memory corruption.

File: src/core/lib/iomgr/tcp_windows.c [hunk 2]
```diff
@@ -401,6 +402,9 @@ grpc_endpoint *grpc_tcp_create(grpc_winsocket *socket, char *peer_string) {
   grpc_closure_init(&tcp->on_read, on_read, tcp);
   grpc_closure_init(&tcp->on_write, on_write, tcp);
   tcp->peer_string = gpr_strdup(peer_string);
+  /* Tell network status tracking code about the new endpoint */
+  grpc_network_status_register_endpoint(&tcp->base);
```
Comment (ctiller) [inline src/core/lib/iomgr/tcp_windows.c]: Where is this unregistered?

**Comment by makdharma** [comment]
@ctiller the build failure seems unrelated and C tests pass. Anything else needed here?

File: src/core/lib/iomgr/network_status_tracker.c [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  }
+  gpr_mu_destroy(&g_endpoint_mutex);
+}
+
+void grpc_network_status_register_endpoint(grpc_endpoint *ep) {
+  if (!g_init_done) {
+    grpc_initialize_network_status_monitor();
+  }
+  gpr_mu_lock(&g_endpoint_mutex);
+  gpr_log(GPR_DEBUG, "Register endpoint %p", ep);
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: Delete log

Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: (or add a trace)

File: src/core/lib/iomgr/network_status_tracker.c [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    head->ep = ep;
+    head->next = prev_head;
+  }
+  gpr_mu_unlock(&g_endpoint_mutex);
+}
+
+void grpc_network_status_unregister_endpoint(grpc_endpoint *ep) {
+  gpr_mu_lock(&g_endpoint_mutex);
+  GPR_ASSERT(head);
+  gpr_log(GPR_DEBUG, "Unregister endpoint %p", ep);
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: ditto

File: src/core/lib/iomgr/network_status_tracker.c [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+void grpc_network_status_shutdown_all_endpoints() {
+  gpr_mu_lock(&g_endpoint_mutex);
+  if (head == NULL) {
+    gpr_mu_unlock(&g_endpoint_mutex);
+    return;
+  }
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+
+  for (endpoint_ll_node *curr = head; curr != NULL; curr = curr->next) {
+    gpr_log(GPR_DEBUG, "Shutting down endpoint %p", curr->ep);
```
Comment (ctiller) [inline src/core/lib/iomgr/network_status_tracker.c]: ditto

**Comment by ctiller** [comment]
LGTM. Put a 'ready to merge' on this once it's green.

**Comment by jtattermusch** [comment]
test this please

**Comment by makdharma** [comment]
test this please

**Comment by jtattermusch** [comment]
Please check that the ruby failure is unrelated https://grpc-testing.appspot.com/job/gRPC_pull_requests/10299/config=opt,language=ruby,platform=linux/console

**Pull request merged**